R/Quosure.R

```diff
@@ -0,0 +1,366 @@
+    lockEnvironment(object@env)
+    object
+  }
+)
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code",
+  signature = c("Quosure", "expression"),
+  function(object, code, name) {
+    code_char <- as.character(code)
+    eval_code(object, code_char, name = name)
+  }
+)
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code",
+  signature = c("Quosure", "language"),
+  function(object, code, name) {
+    code_char <- as.expression(code)
+    eval_code(object, code_char, name = name)
+  }
+)
+
+#' @rdname eval_code
+#' @export
+setMethod(
+  "eval_code",
+  signature = c("Quosure", "ANY"),
+  function(object, code, name) {
+    code_expr <- substitute(code)
+    eval_code(object, code_expr, name = name)
+  }
+)
+
+#' Get object from the `Quosure` environment
+#'
+#' Get object from the `Quosure` environment.
+#' @param object (`Quosure`)
+#' @param var (`character(1)`) name of the variable to pull from the environment.
+#' @name get_var
+#' @export
+setGeneric("get_var", function(object, var) {
+  standardGeneric("get_var")
+})
+
+
+#' @rdname get_var
+#' @export
+setMethod("get_var", signature = c("Quosure", "character"), function(object, var) {
+  get(var, envir = object@env)
+})
+
+#' @param x (`Quosure`)
+#' @param i (`character`) name of the binding in environment (name of the variable)
+#' @param j not used
+#' @param ... not used
+#' @rdname get_var
+#' @export
+setMethod("[[", c("Quosure", "ANY", "missing"), function(x, i, j, ...) {
+  get_var(x, i)
+})
+
+#' Get code from `Quosure`
+#'
+#' @name get_code
+#' @param object (`Quosure`)
+#' @return named `character` with the reproducible code.
+#' @export
+setGeneric("get_code", function(object) {
+  standardGeneric("get_code")
+})
+
+#' @rdname get_code
+#' @export
+setMethod("get_code", signature = "Quosure", function(object) {
+  object@code
+})
+
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `y` has unique code element placed before common element. This means that `y`
+#' in the environment of the `y` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases to be done
+#' @param x (`Quosure`)
+#' @param y (`Quosure`)
+#' @export
+setGeneric("join", function(x, y) {
+  standardGeneric("join")
+})
+
+#' @rdname join
+#' @export
+setMethod("join", signature = c("Quosure", "Quosure"), function(x, y) {
+  join_validation <- check_joinable(x, y)
+
+  # join expressions
+  if (!isTRUE(join_validation)) {
+    stop(join_validation)
+  }
+
+  id_unique <- !y@id %in% x@id
+  x@id <- c(x@id, y@id[id_unique])
+  x@code <- .keep_code_name_unique(x@code, y@code[id_unique])
+
+  # insert (and overwrite) objects from env2 to env
+  x@env <- .copy_env(x@env)
+  lapply(ls(y@env), function(xx) assign(xx, get(xx, y@env), x@env))
```

Comment:
Note that `ls()` has this fun quirk:

![image](https://user-images.githubusercontent.com/15201933/187399117-b62db4d1-5b06-4624-92f0-e6fd25bbed3b.png)

(If you use all.names = TRUE then you will get .random.seed if it exists - we may also want to set sorted = FALSE)